### ISSUE TITLE
[FEATURE REQ] Support required modifier when deserializing for better nullable reference types support

### ISSUE THREAD

**Issue by Methuselah96** [Azure/azure-sdk-for-net#33072]
### Library name

Azure.Data.Tables

### Please describe the feature.

The [docs](https://learn.microsoft.com/en-us/azure/cosmos-db/table/quickstart-dotnet?toc=https%3A%2F%2Flearn.microsoft.com%2Fen-us%2Fazure%2Fstorage%2Ftables%2Ftoc.json&bc=https%3A%2F%2Flearn.microsoft.com%2Fen-us%2Fazure%2Fbread%2Ftoc.json&tabs=azure-cli%2Cwindows#create-an-item) shows creating a table entity by implementing the `ITableEntity` interface, however it requires using the `default!` property initialization "hack" in order to get proper nullable reference types, like so:

```csharp
// C# record type for items in the table
public record Product : ITableEntity
{
    public string RowKey { get; set; } = default!;

    public string PartitionKey { get; set; } = default!;

    public string Name { get; init; } = default!;

    public int Quantity { get; init; }

    public bool Sale { get; init; }

    public ETag ETag { get; set; } = default!;

    public DateTimeOffset? Timestamp { get; set; } = default!;
}
```

I was hoping that the introduction of the `required` attribute in .NET 7 would relieve this particular pain point, like it does for most serialization libraries, however `Azure.Data.Tables` has a `new()` constraint for many of its methods, and so when I try to make the properties `required` I get this error:

```
'Product' cannot satisfy the 'new()' constraint on parameter 'T' in the generic method 'Task<Response<T>> Azure.Data.Tables.TableClient.GetEntityAsync<T>(string, string, IEnumerable<string>, CancellationToken)' because 'Product' has required members
```

Is there a plan for making it so that the `default!` property initialization "hack" is no longer necessary? It seems like the `new()` generic constraint only exists for the case where we're populating a `new()`-able dictionary (usually the `TableEntity` class) and not for deserializing to a custom class that implements `ITableEntity` based on [this code](https://github.com/Azure/azure-sdk-for-net/blob/18eff249806439372a587fae1a87adfe860cdaa6/sdk/tables/Azure.Data.Tables/src/Extensions/DictionaryTableExtensions.cs#L135-L139).

It seems like some options would be:

1. Introduce separate methods for when you're trying to deserialize an entity vs. populating a `new()`-able dictionary, where only the latter would require the `new()` generic constraint.
2. Create the `new()`-able dictionary using `Activator.CreateInstance()`, `FormatterServices.GetUninitializedObject()`, or some other reflection-based approach that doesn't require a generic constraint.

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by christothes** [comment]
Hi @Methuselah96 - Thanks for raising this issue. As it turns out, we didn't really need the `new()` constraint anymore. This has been addressed in #33087
